modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +182,102 @@ private void ensureOpen() {
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
+        // Log at info level every ~5 minutes
+        logForRetry(attempt % 30 == 0 ? Level.INFO : Level.DEBUG, "reading", e);
+        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
+            failures.add(e);
+        }
+        maybeAbort(currentStream);
+        IOUtils.closeWhileHandlingException(currentStream);
+
+        maybeDelaySafely();
+        openStreamWithRetry();
+    }
+
+    private void logForRetry(Level level, String action, Exception e) {
+        if (attempt == 1) {
+            logger.log(
+                level,
+                () -> format(
+                    "failed %s [%s/%s] at offset [%s] with purpose [%s]",
+                    action,
+                    blobStore.bucket(),
+                    blobKey,
+                    start + currentOffset,
+                    purpose.getKey()
+                ),
+                e
+            );
+            return;
+        }
 
         final long meaningfulProgressSize = Math.max(1L, blobStore.bufferSizeInBytes() / 100L);
         final long currentStreamProgress = Math.subtractExact(Math.addExact(start, currentOffset), currentStreamFirstOffset);
         if (currentStreamProgress >= meaningfulProgressSize) {
             failuresAfterMeaningfulProgress += 1;
         }
-        final Supplier<String> messageSupplier = () -> format(
-            """
-                failed reading [%s/%s] at offset [%s]; this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
-                [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
-                retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
-            blobStore.bucket(),
-            blobKey,
-            start + currentOffset,
-            attempt,
-            currentStreamProgress,
-            failuresAfterMeaningfulProgress,
-            maxAttempts
+        logger.log(
+            level,
+            () -> format(
+                """
+                    failed %s [%s/%s] at offset [%s] with purpose [%s]; \
+                    this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
+                    [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
+                    retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
+                action,
+                blobStore.bucket(),
+                blobKey,
+                start + currentOffset,
+                purpose.getKey(),
+                attempt - 1,
+                currentStreamProgress,
+                failuresAfterMeaningfulProgress,
+                attempt - 1 - failuresAfterMeaningfulProgress
```

Comment:
The "maximum number of read attempts" was previously `blobStore.getMaxRetries() + 1` (a constant) and now it seems to me to be variable?

Reply:
Thanks for catching this. I mis-read the message. Pushed cb640128a22ea738f68164a33c927d59da02a2d5